apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -503,30 +516,44 @@ bump_last_alive(S0) ->
     {ok, replies(), session()}.
 replay(ClientInfo, [], Session0 = #{s := S0}) ->
     Streams = emqx_persistent_session_ds_stream_scheduler:find_replay_streams(S0),
-    Session = lists:foldl(
-        fun({_StreamKey, Stream}, SessionAcc) ->
-            replay_batch(Stream, SessionAcc, ClientInfo)
-        end,
-        Session0,
-        Streams
-    ),
+    Session = replay_streams(Session0#{replay => Streams}, ClientInfo),
+    {ok, [], Session}.
+
+replay_streams(Session0 = #{replay := [{_StreamKey, Srs0} | Rest]}, ClientInfo) ->
+    case replay_batch(Srs0, Session0, ClientInfo) of
+        Session = #{} ->
+            replay_streams(Session#{replay := Rest}, ClientInfo);
+        {error, _, _} ->
```

Comment:
Only recoverable errors are actually handled inside `replay_batch/3`. I'll move few pieces of code around to make error handling clearer.

Yet, I think there's a missing TODO as well: what do we do with the session on unrecoverable errors? The simplest approach seems to be to destroy the session and just crash.

Reply:
We can handle it similarly to memq overflow: increase some counters, print some warnings, but let the client proceed.

Reply:
There's a difference though: memq overflow does not affect replays, client just never sees them in the first place. Here, on the other hand, we have some requirements to fulfill so that we won't break the protocol from the client's point of view. Something along the lines of _send special "tombstone" messages in place of those we are no more able to replay_?

Reply:
Indeed, the situations are not exactly the same. memq drops new messages, while DS might drop old ones.
Regarding the tombstone messages... I'm not sure if the clients can properly handle that. It's not part of the standard. We do have the information about the seqnos, though. We may get away from some situations by sending the remaining PUBRELs.

Overall, I agree that for starters it's probably best to drop the session that encounters an unrecoverable error due to retention. BUT I think it has to be done on CONNECT: it would be strange to send a few messages, then suddenly mark the session as expired.